### ISSUE TITLE
Insertion bug

### ISSUE THREAD

**Issue by redbaty** [ralmsdeveloper/EntityFrameworkCore.FirebirdSQL#18]
### The issue

Can't insert properly on Firebird 2.5

### Steps to reproduce

Create a project, add the entity framework dependencies and try to insert multiple objects into a table ( Not tested single ones )

### Error details

It doesn't create an error but it doesn't insert into the database and the SQL output shown on the Output window is not right

### Further technical details

Firebird version: 2.5
EntityFrameworkCore.FirebirdSql version: 2.0.11

**Comment by ralmsdeveloper** [comment]
Hello @redbaty ,
You could open a problem [here ](https://github.com/ralmsdeveloper/EntityFrameworkCore.FirebirdSQL)so we can analyze.

**Comment by ralmsdeveloper** [comment]
Hello @redbaty ,
Could you show us something? any code?

To try to make a repro.

**Comment by redbaty** [comment]
@ralmsdeveloper Hey sorry, I created this just when I was leaving to go home. I'll create one now

**Comment by redbaty** [comment]
@ralmsdeveloper Oh and I've just noted that you can use .EnsureCreated() too, should I create another Issue, to keep track of?

**Comment by ralmsdeveloper** [comment]
Yes, thanks for this!

**Comment by ralmsdeveloper** [comment]
For me it's working EnsureCreated()

**Comment by redbaty** [comment]
@ralmsdeveloper hmm strange, here it gives an error saying that the token BY is unknown, I'll post the stack trace in a bit

**Comment by redbaty** [comment]
@ralmsdeveloper Try to reproduce it using this repo: https://github.com/redbaty/FirebirdTest

**Comment by ralmsdeveloper** [comment]
Make um DownGrade , and test please!

**Comment by redbaty** [comment]
@ralmsdeveloper You mean the package or firebird itself?

**Comment by ralmsdeveloper** [comment]
I will solve some things now, but tomorrow, I will reproduce with your code!

**Comment by redbaty** [comment]
@ralmsdeveloper All right, Boa Noite!

**Comment by ralmsdeveloper** [comment]
Rsss, Boa noite conterrâneo!

**Comment by ralmsdeveloper** [comment]
Hi redbaty,

Thanks for the Issue, please upgrade to 2.0.11.1

**Comment by redbaty** [comment]
@ralmsdeveloper Hey I`ve just updated the same repository with this issue repro. (The insertion bug)

**Comment by ralmsdeveloper** [comment]
Change this:
`public string Url {get; set; }`

Per:
```chsarp
[StringLength (100)]
public string Url {get; set; }
```

There are limitations to internal use of the "BLOCK" command for varchar (8191)

So when you do not tell the maximum size it creates fields to type varchar (8191)

👍

**Comment by redbaty** [comment]
@ralmsdeveloper Hmm, it is still not saving, did it work on your computer?

**Comment by redbaty** [comment]
@ralmsdeveloper Oh wow, that was dumb, I was not adding the toAdd list to the context. It worked thank you so much! :tada:

**Comment by redbaty** [comment]
@ralmsdeveloper Hey just a suggestion though, could you throw an exception when the attribute is not there? Cause then it'll make it easier for others to debug this.

**Comment by ralmsdeveloper** [comment]
Hello @Redbaty,
But this is the expected treatment!
That's right, you just did not pass anything on the EFCore crawler, so by calling SaveChanges (), it will record what's in the crawler!

You can use the `int ret = .... SaveChanges ();`
if the return is greater than 0 recorded!

**Comment by redbaty** [comment]
@ralmsdeveloper Could we add it to the scaffolding process then? Or is it this limitation you were talking about?

**Comment by ralmsdeveloper** [comment]
Not limitation, only you were doing this:
 ```csharp
 for (int i = 0; i < 100; i++)
{
     toAdd.Add(new Blog{Url = Guid.NewGuid().ToString()});
 }
 x.SaveChanges();
```
instead:
 ```csharp
 for (int i = 0; i < 100; i++)
{
     toAdd.Add(new Blog{Url = Guid.NewGuid().ToString()});
 }
 x.Blogs.AddRange(toAdd);  // <<<<---------
 x.SaveChanges();
```
Since there was no information for the crawler then SaveChanges did nothing!

**Comment by redbaty** [comment]
@ralmsdeveloper  I meant these: `[StringLength (100)]`

**Comment by ralmsdeveloper** [comment]
Sorry, I did not see where you referenced this.
But you should do this with Annotation or Fluent API, in OnModelCreating()

Was this class generated by Scaffolding?

**Comment by redbaty** [comment]
From the test project no, but a colleague at work was having some trouble with this, we tried the -d argument when scaffolding but it didn't help, the length attribute was missing

**Comment by redbaty** [comment]
@ralmsdeveloper Actually it does, my bad, but now there's a mistake on the insert statement:

```
INSERT INTO "PRODUTOS" ("CODIGOPRODUTO", "APELIDOPRODUTO", "AREAPRODUTO", "BALANCA", "CEST", "CHKREGISTRO", "CODGRADE", "CODIGOANP", "CODIGOANTERIOR", "CODIGOEMP", "CODIGOINSTRUCAO", "CODNBMSH", "CODORIGINAL", "CODSETOR", "CONTROLAGRADE", "DATACADASTRO", "DESCFISCAL", "DESCORCAMENTO", "DESCRICPRODUTO", "DIASVAL", "FABRICANTEPRODUTO", "FAMILIA", "GRUPOPRODUTO", "IMAGEMPRODUTO", "IMPDATAEMB", "IMPDATAVAL", "LVDESCRICAO", "LVRESUMO", "LVTAG", "MARCAPRODUTO", "MEMOPRODUTO", "NIVEL10PRODUTO", "NIVEL1PRODUTO", "NIVEL2PRODUTO", "NIVEL3PRODUTO", "NIVEL4PRODUTO", "NIVEL5PRODUTO", "NIVEL6PRODUTO", "NIVEL7PRODUTO", "NIVEL8PRODUTO", "NIVEL9PRODUTO", "PERCLOJAONLINE", "PRODUTOLOJAONLINE", "REFERENCIAPRODUTO")

    VALUES (:p0, :p1, :p2, :p3, :p4, :p5, :p6, :p7, :p8, :p9, :p10, :p11, :p12, :p13, :p14, :p15, :p16, :p17, :p18, :p19, :p20, :p21, :p22, :p23, :p24, :p25, :p26, :p27, :p28, :p29, :p30, :p31, :p32, :p33, :p34, :p35, :p36, :p37, :p38, :p39, :p40, :p41, :p42, :p43)END;
```

There's a unnecessary END; at the end.

Ps: this Database is very old and we're moving it to a new platform, that's why there are those horrible column names

**Comment by ralmsdeveloper** [comment]
Open a issue for this, but I believe this is not the full SQL output!

**Comment by redbaty** [comment]
@ralmsdeveloper There is the parameters below, but the error is only occurring at that "END;" we've ran it without it and it worked fine

**Comment by redbaty** [comment]
I'll reopen this issue since this also falls into the "Insertion bug" category

**Comment by redbaty** [comment]
@ralmsdeveloper Hey if you're up to it you can create a slack or something like that so we could debug it faster

**Comment by ralmsdeveloper** [comment]
I'll check that later.
Thanks!

**Comment by redbaty** [comment]
@ralmsdeveloper Thank *YOU*! 😄

**Comment by redbaty** [comment]
@ralmsdeveloper Some progress, I found out that it needs a `;` after the insert statement, now the problem is the affectedrows var on the execute block, it expected 1 but got actually 0

**Comment by redbaty** [comment]
@ralmsdeveloper Okay, got it working. I've added a `suspend;` before the `end;` statement and it worked, but it threw a IndexOutOfRange exception something related about a column name, but the data got inserted so the problem is on C# side.

**Comment by redbaty** [comment]
@ralmsdeveloper Hey, could we really look up that slack / skype / some other instant messager thing ? Cause we need to discuss somethings

**Comment by redbaty** [comment]
This issue seems to have been fixed in the 2.0.11.3 update!
